draft-ietf-quic-load-balancers.md

```diff
@@ -687,23 +687,46 @@ packets that don't include a valid token.
 Both server and service must have access to Universal time, though tight
 synchronization is unnecessary.
 
+The tokens are protected using AES128-GCM AEAD, as explained in
+{{token-protection-with-aead}}.
 All tokens, generated by either the server or retry service, MUST use the
-following format. This format is the cleartext version. On the wire, these
-fields are encrypted using an AES-ECB cipher and the token key. If the token is
-not a multiple of 16 octets, the last block is padded with zeroes.
-
+following format, which includes:
+
+- The client IP address, serialized and used as associated data when
+  computing the AEAD encryption.
+- A token sequence number that must be used only once with a given
+  key. The token sequence number is transmitted in clear text, but
+  is protected as part of the AEAD associated data.
+- A token body, encoding the Original Destination Connection ID,
+  the Retry Source Connection ID, and the Timestamp, optionally
+  followed by server specific Opaque Data.
+
+DISCUSS: shall we specify how the token sequence is formatted? There
+are two main possibilities, a large random number that is statistically
+unique, or the combination of a server ID and a sequence number unique
```

Comment:
I like the ID of prepending the server ID to the sequence number and assigning the Retry Service one or more "server IDs" that keep the sequence number spaces separate, but one problem is that the server ID can be as much as 18 Bytes.

I see three ways forward:
- assign each server a second server ID for these purposes, (blecch)
- just use random numbers
- eliminate the (already very weak) ability of Shared State Retry Services to validate incoming tokens -- they only generate them. Then, each server can do whatever it wants with its self-generated tokens, which should easily be distinguishable by token length or whatever, and the service can use the entire sequence number space without fear of collision.

I'm leaning towards the third.

Reply:
Birthday paradox says with a 64 bit random field you will expect collisions after 2^32 tokens -- and that you have to change the key sooner than that if you want to minimize the risk. If you want the risk of collision below 10E-5 (five nines...) then you should not use the key more than 19 million times.

I should be technically possible to make the identifier 12 bytes long -- that's the normal size of the IV for AES GCM. The same reasoning says that the key should not be use more that 4.0 E+14 times -- 400 trillion times. That may be the way to go. Better than trying to squeeze an 18 bytes ID there.

Or we would have to use something a bit more exotic than AES GCM, but I would rather not go there.

Reply:
One way to manage the key rotation would be to actually configure both entities with a secret and make the retry service and server derive the key from that, so that it can be rotated using some signal that the retry service puts in the token?

Reply:
Or we could just do my third option...